ocaml/xapi/xapi_network.ml

```diff
@@ -330,34 +330,45 @@ let with_networks_attached_for_vm ~__context ?host ~vm f =
     end;
     raise e
 
-(*
-let assert_can_add_purposes ~__context ~network ~current porpoise =
-  if (List.mem current `himn) || (List.mem current `unmanaged) then (
-    info "Network purposes 'himn' and 'unmanaged' are not compatible with any other purpose"
+let assert_can_add_purpose ~__context ~network ~current porpoise =
+  let sop (*string-of-porpoise*) = Record_util.network_purpose_to_string in
+  if (List.mem `himn current) then (
+    info "Existing network purpose %s is not compatible with any other purpose." (sop `himn)
+  );
+  if (List.mem `unmanaged current) then (
+    info "Existing network purpose %s is not compatible with any other purpose." (sop `unmanaged)
   );
-  let assert_no bad_porpoise =
-    Db.see_if_there_is_a_network_with_a bad_porpoise
 
+  let assert_no_net_has_bad_porpoise bads =
+    (* Sadly we can't use Db.Network.get_refs_where because the expression
+     * type doesn't allow searching for a value inside a list. *)
+    Db.Network.get_all ~__context |>
+    List.iter (fun nwk ->
+      Db.Network.get_purposes ~__context ~self:nwk |>
+      List.iter (fun suspect -> if (List.mem suspect bads) then
+        info "Cannot set new network purpose %s when there is a network with purpose %s" (sop porpoise) (sop suspect)
+      )
+    )
+  in
   match porpoise with
   | `himn | `unmanaged -> if current <> [] then (
-      info "Network purposes 'himn' and 'unmanaged' are not compatible with any other purpose"
+      info "New network purpose %s is not compatible with any other purpose" (sop porpoise)
     )
-  | `management -> if (List.mem current `guest) then (
-      info "Network purposes management and guest are incompatible for security reasons."
+  | `management -> if (List.mem `guest current) then (
+      info "Combining network purposes %s and %s is inadvisable for security reasons, but not prevented." (sop `management) (sop `guest)
     )
-  | `guest -> if (List.mem current `management) then (
-      info "Network purposes management and guest are incompatible for security reasons."
+  | `guest -> if (List.mem `management current) then (
+      info "Combining network purposes %s and %s is inadvisable for security reasons, but not prevented." (sop `management) (sop `guest)
     )
   | `storage -> ()
-  | `nbd -> assert_no `insecure_nbd
-  | `insecure_nbd -> assert_no `nbd
-*)
+  | `nbd -> assert_no_net_has_bad_porpoise [`insecure_nbd]
```

Comment:
Please don't bother with anything other than these last two, which we _actually_ need right now.